### PR TITLE
ARROW-16116: [C++] Handle non-nullable fields when reading Parquet

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

**Comment by lidavidm** [comment]
I wasn't able to find a good way to test the int32/int64/byte array decimal cases, perhaps using the Parquet writer API directly to build the file?

**Comment by lidavidm** [comment]
Added a test that directly uses the Parquet writer to generate decimals encoded as binary/int32/int64 to test those paths, too.

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 1]
```diff
@@ -409,22 +441,36 @@ Status TransferDate64(RecordReader* reader, MemoryPool* pool,
...
-                          std::shared_ptr<ChunkedArray>* out) {
+                          bool nullable, std::shared_ptr<ChunkedArray>* out) {
   auto dict_reader = dynamic_cast<DictionaryRecordReader*>(reader);
   DCHECK(dict_reader);
   *out = dict_reader->GetResult();
   if (!logical_value_type->Equals(*(*out)->type())) {
     ARROW_ASSIGN_OR_RAISE(*out, (*out)->View(logical_value_type));
   }
+  if (!nullable) {
+    // Reconstruct each chunk without nulls.
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader_internal.cc:452]: Is this necessary for correctness or is it just an optimization of eliding the null buffer?

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 2]
```diff
@@ -434,14 +480,26 @@ Status TransferBinary(RecordReader* reader, MemoryPool* pool,
...
       // XXX: if a LargeBinary chunk is larger than 2GB, the MSBs of offsets
       // will be lost because they are first created as int32 and then cast to int64.
       ARROW_ASSIGN_OR_RAISE(
-          chunk, ::arrow::compute::Cast(*chunk, logical_value_type, cast_options, &ctx));
+          chunk,
+          ::arrow::compute::Cast(*chunk, logical_type_field->type(), cast_options, &ctx));
     }
   }
-  *out = std::make_shared<ChunkedArray>(chunks, logical_value_type);
+  if (!logical_type_field->nullable()) {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader_internal.cc:491]: Perhaps make a helper function for this operation and refactor?

**Review by pitrou — COMMENTED** [review]
Nice fix, thank you.

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/reader_internal.cc]: Is there a situation where said null buffer would not be all ones?

Comment (pitrou) [reply cpp/src/parquet/arrow/reader_internal.cc]: Hmm, I see, thank you.

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 3]
```diff
@@ -640,43 +696,55 @@ static Status DecimalIntegerTransfer(RecordReader* reader, MemoryPool* pool,
...
   for (size_t i = 0; i < chunks.size(); ++i) {
     std::shared_ptr<Array> chunk_as_decimal;
     auto fn = &DecimalConverter<DecimalArrayType, ParquetType>::ConvertToDecimal;
-    RETURN_NOT_OK(fn(*chunks[i], type, pool, &chunk_as_decimal));
+    RETURN_NOT_OK(fn(*chunks[i], field->type(), pool, &chunk_as_decimal));
     // Replace the chunk, which will hopefully also free memory as we go
     chunks[i] = chunk_as_decimal;
   }
-  *out = std::make_shared<ChunkedArray>(chunks, type);
+  if (!field->nullable()) {
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/parquet/arrow/reader_internal.cc:710]: Ah my bad, fixed now.

**Review by pitrou — APPROVED** [review]
Sorry for the late re-review, LGTM now. Thank you!

**Pull request closed without merging**